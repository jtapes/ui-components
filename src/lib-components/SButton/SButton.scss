@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';

$btn-sizes: (
  'xs': $size--28,
  'sm': $size--32,
  'md': $size--48,
) !default;
$btn-circle-sizes: (
  'xs': $size--32,
  'sm': $size--40,
  'md': $size--48,
) !default;

.s-button {
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 0 $size--16;
  height: $size--48;
  transition-property: background-color, color, border;
  transition-duration: $animate--time;
  cursor: pointer;

  .badge {
    position: absolute;
    right: 0;
    top: 0;
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 0 $size--4;
    height: $size--20;
    min-width: $size--20;
    color: $color--white;
    background-color: $color--danger;
    transform: translateY(-45%);
  }

  &__content {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
    height: 100%;
    white-space: nowrap;
  }

  &__prepend {
    display: flex;
    align-items: center;
    justify-content: center;
    margin-right: $size--8;
    height: 100%;
  }

  &__append {
    display: flex;
    align-items: center;
    justify-content: center;
    margin-left: $size--8;
    height: 100%;
  }

  &:disabled {
    .badge {
      display: none;
    }
  }

  &--rounded {
    border-radius: $size--24;
  }

  &--circle {
    border-radius: 50%;
    padding: 0;

    @each $name, $size in $btn-circle-sizes {
      &.s-button {
        &--#{$name} {
          width: $size;
          height: $size;

          .s-button {
            &__content {
              width: $size;
              height: $size;
            }
          }
        }
      }
    }
  }

  &--default {
    border: 1px solid $color--primary-accent;
    color: $color--white;
    background-color: $color--primary-accent;

    &:hover {
      border: 1px solid $color--primary;
      background-color: $color--primary;
    }

    &:active,
    &.s-button--active {
      border: 1px solid $color--primary-dark;
      background-color: $color--primary-dark;
    }

    &:disabled {
      border: 1px solid $color--background;
      color: $color--border;
      background-color: $color--background;
      cursor: not-allowed;
    }

    &.s-button {
      &--sm {
        @include text-sm;

        border: 1px solid $color--complimentary-bg-light;
        color: $color--primary;
        background-color: $color--complimentary-bg-light;

        &:hover {
          border: 1px solid $color--primary;
          color: $color--white;
          background-color: $color--primary;
        }

        &:active,
        &.s-button--active {
          border: 1px solid $color--background;
          color: $color--primary;
          background-color: $color--background;
        }

        &:disabled {
          border: 1px solid $color--background;
          color: $color--border;
          background-color: $color--background;
          cursor: not-allowed;
        }
      }

      &--inverse {
        border: 1px solid $color--white;
        color: $color--primary;
        background-color: $color--white;

        &:hover {
          border: 1px solid $color--complimentary-bg-light;
          color: $color--primary-accent;
          background-color: $color--complimentary-bg-light;
        }

        &:active,
        &.s-button--active {
          border: 1px solid $color--primary-dark;
          color: $color--white;
          background-color: $color--primary-dark;
        }
      }
    }
  }

  &--outline {
    border: 1px solid $color--border;
    color: $color--primary;
    background-color: transparent;

    &:hover {
      border: 1px solid $color--primary;
      color: $color--white;
      background-color: $color--primary;
    }

    &:active,
    &.s-button--active {
      border: 1px solid $color--primary-dark;
      color: $color--white;
      background-color: $color--primary-dark;
    }

    &:disabled {
      border: 1px solid $color--background;
      color: $color--border;
      background-color: $color--background;
      cursor: not-allowed;
    }

    &.s-button {
      &--inverse {
        border: 1px solid $color--white;
        color: $color--white;
        background-color: transparent;

        &:hover {
          border: 1px solid $color--complimentary-bg-light;
          color: $color--primary-accent;
          background-color: $color--complimentary-bg-light;
        }

        &:active,
        &.s-button--active {
          border: 1px solid $color--primary-dark;
          color: $color--white;
          background-color: $color--primary-dark;
        }
      }
    }
  }

  &--link {
    border: 1px solid transparent;
    color: $color--primary;
    background-color: transparent;

    &:hover {
      border: 1px solid $color--complimentary-bg-light;
      color: $color--primary-accent;
      background-color: $color--complimentary-bg-light;
    }

    &:active,
    &.s-button--active {
      border: 1px solid $color--background;
      color: $color--primary;
      background-color: $color--background;
    }

    &:disabled {
      border: 1px solid transparent;
      color: $color--border;
      background-color: transparent;
      cursor: not-allowed;
    }

    &.s-button {
      &--sm {
        padding: 0;

        &:hover {
          border: 1px solid transparent;
          color: $color--danger;
          background-color: transparent;
        }

        &:active,
        &.s-button--active {
          border: 1px solid transparent;
          color: $color--primary-dark;
          background-color: transparent;
        }

        &:disabled {
          border: 1px solid transparent;
          color: $color--border;
          background-color: transparent;
          cursor: not-allowed;
        }
      }
    }
  }

  &--icon {
    border: 1px solid $color--border;
    color: $color--primary;
    background-color: transparent;

    &:hover {
      border: 1px solid $color--complimentary-bg-light;
      color: $color--primary-accent;
      background-color: $color--complimentary-bg-light;
    }

    &:active,
    &.s-button--active {
      border: 1px solid $color--background;
      color: $color--primary;
      background-color: $color--background;
    }

    &:disabled {
      border: 1px solid $color--background;
      color: $color--border;
      background-color: $color--background;
      cursor: not-allowed;
    }
  }

  &--favorite {
    border: 1px solid transparent;
    color: $color--font-secondary;
    background-color: transparent;

    &:hover {
      border: 1px solid $color--complimentary-bg-light;
      color: $color--primary-accent;
      background-color: $color--complimentary-bg-light;
    }

    &:active {
      border: 1px solid $color--primary-dark;
      color: $color--white;
      background-color: $color--primary-dark;
    }

    &.s-button--active {
      border: 1px solid transparent;
      color: $color--danger;
      background-color: transparent;

      &:hover {
        border: 1px solid $color--border;
        color: $color--danger;
        background-color: transparent;
      }

      &:active {
        border: 1px solid $color--primary-dark;
        color: $color--white;
        background-color: $color--primary-dark;
      }
    }
  }

  &--grid {
    border: 1px solid $color--border;
    color: $color--primary;
    background-color: $color--background;

    &:hover {
      color: $color--primary-accent;
      background-color: $color--complimentary-bg-light;
    }

    &:active,
    &.s-button--active {
      color: $color--primary;
      background-color: $color--white;
    }

    &:disabled {
      border: 1px solid $color--background;
      color: $color--border;
      background-color: $color--background;
      cursor: not-allowed;
    }
  }

  &--clear {
    margin: 0;
    padding: 0;
    color: $color--font-muted;
    background-color: transparent;

    &:hover {
      color: $color--primary;
    }
  }

  &--search {
    margin: 0;
    border: 1px solid $color--background;
    padding: 0;
    color: $color--primary;
    background-color: $color--background;

    &:hover {
      border: 1px solid $color--primary-dark;
      color: $color--white;
      background-color: $color--primary-dark;
    }

    &:active {
      border: 1px solid $color--primary-dark;
      background-color: $color--primary-dark;
    }
  }

  & + & {
    margin-left: $size--20;
  }

  @each $name, $size in $btn-sizes {
    &--#{$name} {
      height: $size;
    }
  }
}
