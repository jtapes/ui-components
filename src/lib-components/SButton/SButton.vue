<template>
  <button
    class="s-button"
    :disabled="state === 'disabled'"
    :class="[
      `s-button--${mode}`,
      `s-button--${size}`,
      {
        's-button--inverse': inverse,
        's-button--rounded': rounded,
        's-button--circle': mode === 'icon' || mode === 'favorite' || circle,
        's-button--active': state === 'active',
      },
    ]"
    :type="type"
    v-on="$listeners"
  >
    <div v-if="$slots.prepend" class="s-button__prepend">
      <!-- @slot Слот для вставки перед контентом кнопки -->
      <slot name="prepend" />
    </div>

    <div class="s-button__content" :class="contentClass">
      <!-- @slot Слот для контента кнопки -->
      <slot />
    </div>

    <span v-if="badge" class="badge">{{ badge }}</span>

    <div v-if="$slots.append" class="s-button__append">
      <!-- @slot Слот для вставки после контента кнопки -->
      <slot name="append" />
    </div>
  </button>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'
import SvgIcon from '@/lib-components/SvgIcon/SvgIcon.vue'

@Component({
  components: {
    SvgIcon,
  }
})
export default class SButton extends Vue {
  /** Свойство типа кнопки */
  @Prop({ type: String, default: 'button' })
  type!: string

  /** Свойство режима кнопки */
  @Prop({ type: String, default: 'default' })
  mode!: string

  /** Свойство округления кнопки */
  @Prop({ type: Boolean, default: true })
  rounded!: boolean

  /** Форма кнопки (круглая) */
  @Prop({ type: Boolean, default: false })
  circle!: boolean

  /** Размер кнопки ("xs", "sm", "md") */
  @Prop({ type: String, default: 'md' })
  size!: string

  /** Badge - значок с количеством чего-либо  */
  @Prop(Number)
  badge?: number

  /** Инверсированная кнопка */
  @Prop({ type: Boolean, default: false })
  inverse!: boolean

  /** Состояние кнопки ("disabled", "active") */
  @Prop({ type: String, default: 'default' })
  state!: string

  /** Дополнительные классы для контента кнопки */
  @Prop(String)
  contentClass?: string
}
</script>
