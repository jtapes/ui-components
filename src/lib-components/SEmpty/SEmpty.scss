@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';

.s-empty {
  display: flex;
  flex-direction: column;
  align-items: center;
  margin-top: $size--60;

  &__description {
    margin: $size--24 0 $size--40;
  }
}
