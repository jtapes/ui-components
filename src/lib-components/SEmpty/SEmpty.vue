<template>
  <div class="s-empty">
    <div class="s-empty__title">
      <!-- @slot Слот для заголовка -->
      <slot name="title">
        <span class="text-xl color-font-muted">
          {{ title }}
        </span>
      </slot>
    </div>
    <div v-if="description || $slots.description" class="s-empty__description">
      <!-- @slot Слот для описания -->
      <slot name="description">
        {{ description }}
      </slot>
    </div>
    <div v-if="$slots.content" class="s-empty__content">
      <!-- @slot Слот для контента -->
      <slot name="content" />
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'

/** Компонент для отображения информации, когда данных нет */
@Component
export default class SEmpty extends Vue {
  /** Заголовок */
  @Prop({ type: String, default: 'Нет данных' })
  title!: string

  /** Описание */
  @Prop(String)
  description?: string
}
</script>
