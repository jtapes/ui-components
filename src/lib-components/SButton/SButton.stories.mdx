import { action } from '@storybook/addon-actions'
import { config, withDesign } from 'storybook-addon-designs'
import { withTests } from '@storybook/addon-jest'
import {
  Meta,
  Story,
  Source,
  Canvas,
  ArgsTable,
  Description,
} from '@storybook/addon-docs'
import SButton from '@/lib-components/SButton/SButton.vue'
import SvgIcon from '@/lib-components/SvgIcon/SvgIcon.vue'
import { SvgMdxArgType } from '@/storybook/components/icon/SvgIconArgType'

import results from '/.jest-test-results.json'

<Meta
  title="DESIGN SYSTEM/Atom/Button"
  component={SButton}
  decorators={[withDesign, withTests({ results })]}
  argTypes={{
    content: { control: 'text', defaultValue: 'В корзину' },
    state: {
      control: { type: 'select' },
      options: [
        'disabled',
        'active',
        'default'
      ]
    },
    append: SvgMdxArgType,
    prepend: SvgMdxArgType,
    icon: SvgMdxArgType,
    type: {
      options: ['button', 'submit', 'reset'],
      control: { type: 'select', defaultValue: 'button' },
    },
    mode: {
      options: ['default', 'outline', 'link', 'icon', 'favorite', 'grid'],
      control: { type: 'select', defaultValue: 'default' },
    },
    circle: { control: 'boolean', defaultValue: false },
    size: {
      options: [
        'xs',
        'sm',
        'md',
      ],
      control: { type: 'select' },
      defaultValue: 'md',
    },
    rounded: { control: 'boolean', defaultValue: true },
    badge: { control: 'text' },
    inverse: { control: 'boolean', defaultValue: false },
    onClick: { action: 'clicked' },
  }}
  parameters={{
    design: config({
      type: 'figma',
      url:
        'https://www.figma.com/file/IIeFkxNoHGcaqNHlVeNrhW/MVP-3-Market-UI-Kit?node-id=3449%3A2819',
    }),
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>

export const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { SButton, SvgIcon },
  template: `
    <s-button
      :state="state"
      :content="content"
      :append="append"
      :prepend="prepend"
      :icon="icon"
      :type="type"
      :mode="mode"
      :circle="circle"
      :size="size"
      :rounded="rounded"
      :badge="badge"
      :inverse="inverse"
      :onClick="clickme">
      <template v-if="icon">
        <svg-icon :name="icon" class="icon-32" />
      </template>
      <template v-else>
        <span>{{ content }}</span>
      </template>
      <template #prepend v-if="prepend">
        <svg-icon :name="prepend" class="icon-32" />
      </template>
      <template #append v-if="append">
        <svg-icon :name="append" :class="append === 'plus-16' ? 'icon-16' : 'icon-32'" />
      </template>
    </s-button>
  `,
  methods: {
    clickme() {
      action('button-click')()
    },
  },
})

export const InverseTemplate = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { SButton, SvgIcon },
  template: `
    <div style="background-color: #002F6C; padding: 40px;">
      <s-button
      :state="state"
      :content="content"
      :append="append"
      :prepend="prepend"
      :icon="icon"
      :type="type"
      :mode="mode"
      :circle="circle"
      :size="size"
      :rounded="rounded"
      :badge="badge"
      :inverse="inverse"
      :onClick="clickme">
        <template v-if="icon">
          <svg-icon :name="icon" class="icon-32" />
        </template>
        <template v-else>
          <span>{{ content }}</span>
        </template>
        <template #prepend v-if="prepend">
          <svg-icon :name="prepend" class="icon-32" />
        </template>
        <template #append v-if="append">
          <svg-icon :name="append" class="icon-32" />
        </template>
      </s-button>
    </div>
  `,
  methods: {
    clickme() {
      action('button-click')()
    },
  },
})

# Кнопка

<Description />

<Canvas withToolbar withSource="none">
  <Story name="SButton (default)">
    {Template.bind({})}
  </Story>
  <Story
    name="SButton (outline)"
    args={{
      mode: 'outline',
      append: 'arrow-24-32'
    }}
  >
    {Template.bind({})}
  </Story>
  <Story name="SButton (default inverse)" args={{ inverse: true }}>
    {InverseTemplate.bind({})}
  </Story>
  <Story
    name="SButton (outline inverse)"
    args={{
      mode: 'outline',
      append: 'arrow-24-32',
      inverse: true
    }}
  >
    {InverseTemplate.bind({})}
  </Story>
  <Story
    name="SButton (link)"
    args={{
      mode: 'link',
      content: 'Кнопка',
      append: 'arrow-24-32'
    }}
  >
    {Template.bind({})}
  </Story>
  <Story
    name="SButton (icon)"
    args={{
      mode: 'icon',
      append: null,
      icon: 'heart-24-32'
    }}
  >
    {Template.bind({})}
  </Story>
  <Story
    name="SButton (favorite)"
    args={{
      mode: 'favorite',
      append: null,
      icon: 'heart-24-32',
    }}
  >
    {Template.bind({})}
  </Story>
  <Story
    name="SButton (default small)"
    args={{
      content: 'Подписаться',
      append: null,
      size: 'sm'
    }}
  >
    {Template.bind({})}
  </Story>
  <Story
    name="SButton (link small)"
    args={{
      mode: 'link',
      content: 'Показать ещё',
      append: 'plus-16',
      size: 'sm'
    }}
  >
    {Template.bind({})}
  </Story>
  <Story
    name="SButton (grid)"
    args={{
      rounded: false,
      mode: 'grid',
      content: 'Очистить все фильтры',
      append: 'close-24-32'
    }}
  >
    {Template.bind({})}
  </Story>
  <Story
    name="SButton (add to basket)"
    args={{
      append: 'plus-24-32',
    }}
  >
    {Template.bind({})}
  </Story>
  <Story
    name="SButton (show more)"
    args={{
      content: 'Показать ещё',
      mode: 'outline',
      append: 'plus-24-32',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

## Пример

<Source
  language="html"
  code={`
<s-button mode="icon" @click.native="clickme" circle size="sm" :badge="1">
  <template #prepend>
    <svg-icon name="burger-24-32" />
  </template>
  <svg-icon name="heart-24-32" />
  <template #append>
    <svg-icon name="plus-24-32" />
  </template>
</s-button>
  `}
/>

<ArgsTable />
