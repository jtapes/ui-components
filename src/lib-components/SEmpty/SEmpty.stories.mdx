import { config, withDesign } from 'storybook-addon-designs'
import { withTests } from '@storybook/addon-jest';
import {
  Meta,
  Story,
  ArgsTable,
  Description,
  Source,
  Preview,
} from '@storybook/addon-docs'
import results from '/.jest-test-results.json';
import SEmpty from '@/lib-components/SEmpty/SEmpty.vue'
import SButton from '@/lib-components/SButton/SButton.vue'
import SvgIcon from '@/lib-components/SvgIcon/SvgIcon.vue'

<Meta
  title="DESIGN SYSTEM/Atom/Empty"
  component={SEmpty}
  decorators={[withDesign, withTests({results})]}
  argTypes={{
    title: { control: 'text' },
    description: { control: 'text' },
  }}
  parameters={{
    design: config({
      type: 'figma',
      url:
        'https://www.figma.com/file/42Xh2qI3qvq2CRkc2HJlwA/MVP-3-Market-Catalog-%E2%80%A2-Product-Page?node-id=1902%3A2603',
    }),
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>

# Компонент для отображения информации, когда данных нет

<Description of={SEmpty} />

export const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { SEmpty, SButton, SvgIcon },
  template:
    `<div>
      <s-empty :title="title" :description="description">
        <template #content>
          <s-button mode="outline">
            Очистить фильтр
            <template #append>
              <svg-icon name="close-24-32" class="icon-32" />
            </template>
          </s-button>
        </template>
      </s-empty>
    </div>`,
})

<Preview withToolbar withSource="none">
  <Story name="SEmpty" args={{ title: 'Нет подходящих товаров', description: 'Попробуйте изменить условия в фильтрах' }}>
    {Template.bind({})}
  </Story>
</Preview>

## Пример

<Source
  language="html"
  code={`
<s-empty
  v-if="!products.length"
  title="Нет подходящих товаров"
  description="Попробуйте изменить условия в фильтрах"
>
  <template #content>
    <s-button mode="outline">
      Очистить фильтр
      <template #append>
        <svg-icon name="close-24-32" class="icon-32" />
      </template>
    </s-button>
  </template>
</s-empty>
  `}
/>

<ArgsTable of={SEmpty} />
