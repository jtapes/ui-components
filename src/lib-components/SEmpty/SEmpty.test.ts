import { mount } from '@vue/test-utils'
import SEmpty from '@/lib-components/SEmpty/SEmpty.vue'
import SButton from '@/lib-components/SButton/SButton.vue'

describe('SEmpty', () => {
  test('Is Vue snapshot', () => {
    const wrapper = mount(SEmpty, {
      propsData: {
        title: 'Нет подходящих товаров',
        description: 'Попробуйте изменить условия в фильтрах',
      },
      stubs: {
        's-button': SButton,
      },
      slots: {
        content: `
          <s-button mode="outline">
            Очистить фильтр
            <template #append>
              <svg-icon name="close-24-32" class="icon-32" />
            </template>
          </s-button>`,
      },
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
