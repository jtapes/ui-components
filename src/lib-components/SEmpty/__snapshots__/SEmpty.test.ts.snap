// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SEmpty Is Vue snapshot 1`] = `
<div
  class="s-empty"
>
  <div
    class="s-empty__title"
  >
    <span
      class="text-xl color-font-muted"
    >
      
        Нет подходящих товаров
      
    </span>
  </div>
   
  <div
    class="s-empty__description"
  >
    
      Попробуйте изменить условия в фильтрах
    
  </div>
   
  <div
    class="s-empty__content"
  >
    <button
      class="s-button s-button--outline s-button--md s-button--rounded"
      type="button"
    >
      <!---->
       
      <div
        class="s-button__content"
      >
        
            Очистить фильтр
            
      </div>
       
      <!---->
       
      <div
        class="s-button__append"
      >
        <div
          class="mock-component icon-32"
          name="close-24-32"
        />
      </div>
    </button>
  </div>
</div>
`;
